processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java

```diff
@@ -395,11 +498,20 @@ public Row apply(@Nullable Map.Entry<ByteBuffer, Integer> input)
                   ByteBuffer keyBuffer = input.getKey().duplicate();
                   for (int i = 0; i < dimensions.size(); ++i) {
                     final DimensionSelector dimSelector = dimensions.get(i);
-                    final DimExtractionFn fn = dimensionSpecs.get(i).getDimExtractionFn();
+                    final DimensionSpec dimensionSpec = dimensionSpecs.get(i);
+                    final ExtractionFn extractionFn = dimensionSpec.getExtractionFn();
                     final int dimVal = keyBuffer.getInt();
                     if (dimSelector.getValueCardinality() != dimVal) {
-                      if (fn != null) {
-                        theEvent.put(dimNames.get(i), fn.apply(dimSelector.lookupName(dimVal)));
+                      if (extractionFn != null) {
+                        final String value;
+                        // extraction functions are applied at the dimSelector level for time
```

Comment:
Can you explain this optimization a bit more?  Also, have you benchmarked if it actually helps?  It seems like it is adding extra branch complexity in the code, which can sometimes be counter productive.